{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "from cobra.medium import minimal_medium\n",
    "from cobra.flux_analysis import (\n",
    "    single_gene_deletion, single_reaction_deletion, double_gene_deletion,\n",
    "    double_reaction_deletion)\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''load the base model for REL606.'''\n",
    "base_model = cobra.io.read_sbml_model('../metabolic-modeling/iECB_1328.xml.gz')\n",
    "max_growth = base_model.slim_optimize()\n",
    "base_model.medium = minimal_medium(base_model, max_growth)\n",
    "\n",
    "''' set glucose to 1.0. This is to set relative molarity of citrate and glucose in the media properly,\n",
    "so that growth yields of Cit- on DM25 and Cit+ on DM0 and DM25 are \"apples to apples\". '''\n",
    "citminus_medium = base_model.medium\n",
    "citminus_medium['EX_glc__D_e'] = 1.0\n",
    "base_model.medium = citminus_medium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'nad_c + oaa_c --> co2_e + nadh_c + pyr_c'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "''' add this reaction to maeA: nad[c] + oaa[c] ->pyr[c] + nadh[c].\n",
    "    This reaction is the difference between base_model and maeA_model.'''\n",
    "\n",
    "maeA_side_reaction = cobra.Reaction('MAEA')\n",
    "maeA_side_reaction.name = 'maeA_side_rxn '\n",
    "maeA_side_reaction.subsystem = 'Anaplerotic Reactions'\n",
    "maeA_side_reaction.lower_bound = 0.  # This is the default\n",
    "maeA_side_reaction.upper_bound = 1000.  # This is the default\n",
    "\n",
    "maeA_model = base_model.copy()\n",
    "maeA_model.add_reaction(maeA_side_reaction)\n",
    "\n",
    "''' cite this paper: The PEP–pyruvate–oxaloacetate node as the switch point for carbon flux distribution in bacteria\n",
    "    by Uwe Sauer and Bernhard J. Eikmanns.'''\n",
    "\n",
    "maeA_side_reaction.add_metabolites({\n",
    "    'nad_c': -1.0,\n",
    "    'oaa_c': -1.0,\n",
    "    'pyr_c': 1.0,\n",
    "    'nadh_c': 1.0,\n",
    "    'co2_e': 1.0,\n",
    "})\n",
    "\n",
    "maeA_side_reaction.reaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'EX_ca2_e': 0.005113800273256523,\n",
       " 'EX_cl_e': 0.005113800273256523,\n",
       " 'EX_cobalt2_e': 2.4561960966649965e-05,\n",
       " 'EX_cu2_e': 0.000696577213014193,\n",
       " 'EX_fe2_e': 0.015779586200559713,\n",
       " 'EX_glc__D_e': 1.0,\n",
       " 'EX_k_e': 0.19177291387853226,\n",
       " 'EX_mg2_e': 0.008523000455427539,\n",
       " 'EX_mn2_e': 0.0006788926011182051,\n",
       " 'EX_mobd_e': 0.00012673971858791382,\n",
       " 'EX_nh4_e': 10.611576699826303,\n",
       " 'EX_ni2_e': 0.0003173405356891175,\n",
       " 'EX_o2_e': 17.57432917781992,\n",
       " 'EX_pi_e': 0.9477291587688313,\n",
       " 'EX_so4_e': 0.24779088701595114,\n",
       " 'EX_zn2_e': 0.0003350251475851055}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "''' TODO: USE CITROBACTER AS A CONTROL\n",
    "    PROBLEM: KBase and BiGG database IDs are not compatible!\n",
    "    Have to figure out how to do the mapping.. so skip this for now. \n",
    "    load the curated Citrobacter model (Cuepas et al. 2016)\n",
    "    use this model as a positive control to compare flux distribution.'''\n",
    "citrobacter_model = cobra.io.read_sbml_model('../metabolic-modeling/C.sedlakii_MAP_gapfill.sbml')\n",
    "citro_max_growth = citrobacter_model.slim_optimize()\n",
    "''' give citrobacter the same media as base'''\n",
    "base_model.medium\n",
    "#citrobacter_model.medium = minimal_medium(base_model, max_growth)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "''' NOTE: This code uses BiGG model IDs (BiGG namespace).'''\n",
    "def generate_models(basemodel):\n",
    "    ''' Cit- model growing on glucose (no citrate flux)'''\n",
    "    glucose_model = basemodel.copy()\n",
    "    ''' Cit+ model on DM0 with citrate flux'''\n",
    "    DM0citrate_model = basemodel.copy()\n",
    "    ## set glucose to zero, and citrate to 10.0.\n",
    "    DM0 = DM0citrate_model.medium\n",
    "    DM0['EX_glc__D_e'] = 0.0\n",
    "    DM0['EX_cit_e'] = 10.0\n",
    "    DM0citrate_model.medium = DM0\n",
    "\n",
    "    assert 'EX_cit_e' in DM0citrate_model.medium\n",
    "    assert 'EX_glc__D_e' not in DM0citrate_model.medium\n",
    "    assert 'EX_glc__D_e' in glucose_model.medium\n",
    "    assert 'EX_cit_e' not in glucose_model.medium\n",
    "\n",
    "    ''' Cit+ model on DM25 with citrate flux'''\n",
    "    DM25citrate_model = DM0citrate_model.copy()\n",
    "    DM25 = DM25citrate_model.medium\n",
    "    DM25['EX_glc__D_e'] = 1.0\n",
    "    DM25['EX_cit_e'] = 10.0\n",
    "    DM25citrate_model.medium = DM25\n",
    "\n",
    "    ''' now make the fermentation (no O2 influx) models.'''\n",
    "    glucose_fermentation_model = glucose_model.copy()\n",
    "    noO2_media = glucose_fermentation_model.medium\n",
    "    noO2_media['EX_o2_e'] = 0.0\n",
    "    glucose_fermentation_model.medium = noO2_media\n",
    "    \n",
    "    DM0citrate_fermentation_model = DM0citrate_model.copy()\n",
    "    DM0_noO2_media = DM0citrate_fermentation_model.medium\n",
    "    DM0_noO2_media['EX_o2_e'] = 0.0\n",
    "    DM0citrate_fermentation_model.medium = DM0_noO2_media\n",
    "    \n",
    "    DM25citrate_fermentation_model = DM25citrate_model.copy()\n",
    "    DM25_noO2_media = DM25citrate_fermentation_model.medium\n",
    "    DM25_noO2_media['EX_o2_e'] = 0.0\n",
    "    DM25citrate_fermentation_model.medium = DM25_noO2_media\n",
    "    \n",
    "    \n",
    "    return (glucose_model, DM0citrate_model, DM25citrate_model,glucose_fermentation_model, DM0citrate_fermentation_model, DM25citrate_fermentation_model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "old_cit_minus_model, old_DM0_cit_plus_model, old_DM25_cit_plus_model,old_cit_minus_ferment_model, old_DM0_cit_plus_ferment_model, old_DM25_cit_plus_ferment_model = generate_models(base_model)\n",
    "\n",
    "new_cit_minus_model, new_DM0_cit_plus_model, new_DM25_cit_plus_model,new_cit_minus_ferment_model, new_DM0_cit_plus_ferment_model, new_DM25_cit_plus_ferment_model = generate_models(maeA_model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "old_cit_minus_solution = old_cit_minus_model.optimize()\n",
    "old_DM0_cit_plus_solution = old_DM0_cit_plus_model.optimize()\n",
    "old_DM25_cit_plus_solution = old_DM25_cit_plus_model.optimize()\n",
    "\n",
    "new_cit_minus_solution = new_cit_minus_model.optimize()\n",
    "new_DM0_cit_plus_solution = new_DM0_cit_plus_model.optimize()\n",
    "new_DM25_cit_plus_solution = new_DM25_cit_plus_model.optimize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cobra/util/solver.py:416 \u001b[1;31mUserWarning\u001b[0m: solver status is 'infeasible'\n"
     ]
    }
   ],
   "source": [
    "old_cit_minus_ferment_solution = old_cit_minus_ferment_model.optimize()\n",
    "old_DM0_cit_plus_ferment_solution = old_DM0_cit_plus_ferment_model.optimize()\n",
    "old_DM25_cit_plus_ferment_solution = old_DM25_cit_plus_ferment_model.optimize()\n",
    "\n",
    "new_cit_minus_ferment_solution = new_cit_minus_ferment_model.optimize()\n",
    "new_DM0_cit_plus_ferment_solution = new_DM0_cit_plus_ferment_model.optimize()\n",
    "new_DM25_cit_plus_ferment_solution = new_DM25_cit_plus_ferment_model.optimize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n",
      "-3.539511906652807e-29\n",
      "1.1132132731977292e-29\n"
     ]
    }
   ],
   "source": [
    "''' all fermentation solutions are dead.'''\n",
    "print(new_cit_minus_ferment_solution.objective_value)\n",
    "print(new_DM0_cit_plus_ferment_solution.objective_value)\n",
    "print(new_DM25_cit_plus_ferment_solution.objective_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.08985434064378761\n",
      "0.5508334988566986\n",
      "0.6563960816941321\n"
     ]
    }
   ],
   "source": [
    "print(new_cit_minus_solution.objective_value)\n",
    "print(new_DM0_cit_plus_solution.objective_value)\n",
    "print(new_DM25_cit_plus_solution.objective_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_metabolites(mymodel):\n",
    "    print(mymodel.metabolites.cit_c.summary())\n",
    "    print(mymodel.metabolites.acon_C_c.summary())\n",
    "    print(mymodel.metabolites.icit_c.summary())\n",
    "    print(mymodel.metabolites.accoa_c.summary())\n",
    "    print(mymodel.metabolites.pyr_c.summary())\n",
    "    print(mymodel.metabolites.mal__L_c.summary())\n",
    "    print(mymodel.metabolites.oaa_c.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n",
      "0.0\n",
      "0.0\n"
     ]
    }
   ],
   "source": [
    "''' no predicted flux through Citrate Lyase.'''\n",
    "print(new_cit_minus_solution['CITL'])\n",
    "print(new_DM0_cit_plus_solution['CITL'])\n",
    "print(new_DM25_cit_plus_solution['CITL'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n",
      "0.0\n",
      "0.0\n"
     ]
    }
   ],
   "source": [
    "print(old_cit_minus_solution['ME1'])\n",
    "print(old_DM0_cit_plus_solution['ME1'])\n",
    "print(old_DM25_cit_plus_solution['ME1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n",
      "6.187856834140581\n",
      "6.791459892307894\n"
     ]
    }
   ],
   "source": [
    "print(old_cit_minus_solution['ME2'])\n",
    "print(old_DM0_cit_plus_solution['ME2'])\n",
    "print(old_DM25_cit_plus_solution['ME2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n",
      "0.0\n",
      "0.0\n"
     ]
    }
   ],
   "source": [
    "print(new_cit_minus_solution['ME1'])\n",
    "print(new_DM0_cit_plus_solution['ME1'])\n",
    "print(new_DM25_cit_plus_solution['ME1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n",
      "6.187856834140581\n",
      "6.791459892307894\n"
     ]
    }
   ],
   "source": [
    "print(old_cit_minus_solution['ME2'])\n",
    "print(old_DM0_cit_plus_solution['ME2'])\n",
    "print(old_DM25_cit_plus_solution['ME2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.5751379889373442\n",
      "17.65997835679671\n",
      "13.80691340788684\n"
     ]
    }
   ],
   "source": [
    "print(new_cit_minus_solution['MAEA'])\n",
    "print(new_DM0_cit_plus_solution['MAEA'])\n",
    "print(new_DM25_cit_plus_solution['MAEA'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PRODUCING REACTIONS -- Nicotinamide adenine dinucleotide - reduced (nadh_c)\n",
      "---------------------------------------------------------------------------\n",
      "%      FLUX  RXN ID    REACTION\n",
      "---  ------  --------  --------------------------------------------------\n",
      "29%   1.6    GAPD      g3p_c + nad_c + pi_c <=> 13dpg_c + h_c + nadh_c\n",
      "29%   1.58   MAEA      nad_c + oaa_c --> co2_e + nadh_c + pyr_c\n",
      "24%   1.3    PDH       coa_c + nad_c + pyr_c --> accoa_c + co2_c + nadh_c\n",
      "15%   0.826  MDH       mal__L_c + nad_c <=> h_c + nadh_c + oaa_c\n",
      "3%    0.149  PGCD      3pg_c + nad_c --> 3php_c + h_c + nadh_c\n",
      "\n",
      "CONSUMING REACTIONS -- Nicotinamide adenine dinucleotide - reduced (nadh_c)\n",
      "---------------------------------------------------------------------------\n",
      "%      FLUX  RXN ID    REACTION\n",
      "---  ------  --------  --------------------------------------------------\n",
      "98%   5.32   NADH16pp  4.0 h_c + nadh_c + q8_c --> 3.0 h_p + nad_c + q...\n",
      "None\n",
      "************************\n",
      "PRODUCING REACTIONS -- Nicotinamide adenine dinucleotide - reduced (nadh_c)\n",
      "---------------------------------------------------------------------------\n",
      "%      FLUX  RXN ID    REACTION\n",
      "---  ------  --------  --------------------------------------------------\n",
      "47%  18.3    MDH       mal__L_c + nad_c <=> h_c + nadh_c + oaa_c\n",
      "35%  13.8    MAEA      nad_c + oaa_c --> co2_e + nadh_c + pyr_c\n",
      "13%   5.08   AKGDH     akg_c + coa_c + nad_c --> co2_c + nadh_c + succ...\n",
      "3%    1.13   PGCD      3pg_c + nad_c --> 3php_c + h_c + nadh_c\n",
      "1%    0.42   GAPD      g3p_c + nad_c + pi_c <=> 13dpg_c + h_c + nadh_c\n",
      "1%    0.296  IPMD      3c2hmp_c + nad_c --> 3c4mop_c + h_c + nadh_c\n",
      "\n",
      "CONSUMING REACTIONS -- Nicotinamide adenine dinucleotide - reduced (nadh_c)\n",
      "---------------------------------------------------------------------------\n",
      "%      FLUX  RXN ID    REACTION\n",
      "---  ------  --------  --------------------------------------------------\n",
      "58%  22.8    NADH16pp  4.0 h_c + nadh_c + q8_c --> 3.0 h_p + nad_c + q...\n",
      "38%  15      THD2pp    2.0 h_p + nadh_c + nadp_c --> 2.0 h_c + nad_c +...\n",
      "1%    0.234  HACD1     aacoa_c + h_c + nadh_c <=> 3hbcoa_c + nad_c\n",
      "1%    0.234  HACD2     3ohcoa_c + h_c + nadh_c <=> 3hhcoa_c + nad_c\n",
      "1%    0.234  HACD3     3oocoa_c + h_c + nadh_c <=> 3hocoa_c + nad_c\n",
      "1%    0.234  HACD4     3odcoa_c + h_c + nadh_c <=> 3hdcoa_c + nad_c\n",
      "1%    0.234  HACD5     3oddcoa_c + h_c + nadh_c <=> 3hddcoa_c + nad_c\n",
      "None\n",
      "************************\n",
      "PRODUCING REACTIONS -- Nicotinamide adenine dinucleotide - reduced (nadh_c)\n",
      "---------------------------------------------------------------------------\n",
      "%      FLUX  RXN ID    REACTION\n",
      "---  ------  --------  --------------------------------------------------\n",
      "55%  22.7    MDH       mal__L_c + nad_c <=> h_c + nadh_c + oaa_c\n",
      "42%  17.7    MAEA      nad_c + oaa_c --> co2_e + nadh_c + pyr_c\n",
      "2%    0.947  PGCD      3pg_c + nad_c --> 3php_c + h_c + nadh_c\n",
      "1%    0.248  IPMD      3c2hmp_c + nad_c --> 3c4mop_c + h_c + nadh_c\n",
      "\n",
      "CONSUMING REACTIONS -- Nicotinamide adenine dinucleotide - reduced (nadh_c)\n",
      "---------------------------------------------------------------------------\n",
      "%      FLUX  RXN ID    REACTION\n",
      "---  ------  --------  --------------------------------------------------\n",
      "56%  23.3    NADH16pp  4.0 h_c + nadh_c + q8_c --> 3.0 h_p + nad_c + q...\n",
      "38%  15.9    THD2pp    2.0 h_p + nadh_c + nadp_c --> 2.0 h_c + nad_c +...\n",
      "3%    1.33   GAPD      g3p_c + nad_c + pi_c <=> 13dpg_c + h_c + nadh_c\n",
      "None\n",
      "************************\n",
      "************************\n",
      "************************\n",
      "PRODUCING REACTIONS -- Nicotinamide adenine dinucleotide (nad_c)\n",
      "----------------------------------------------------------------\n",
      "%       FLUX  RXN ID    REACTION\n",
      "----  ------  --------  --------------------------------------------------\n",
      "100%   5.32   NADH16pp  4.0 h_c + nadh_c + q8_c --> 3.0 h_p + nad_c + q...\n",
      "\n",
      "CONSUMING REACTIONS -- Nicotinamide adenine dinucleotide (nad_c)\n",
      "----------------------------------------------------------------\n",
      "%       FLUX  RXN ID    REACTION\n",
      "----  ------  --------  --------------------------------------------------\n",
      "30%    1.6    GAPD      g3p_c + nad_c + pi_c <=> 13dpg_c + h_c + nadh_c\n",
      "30%    1.58   MAEA      nad_c + oaa_c --> co2_e + nadh_c + pyr_c\n",
      "24%    1.3    PDH       coa_c + nad_c + pyr_c --> accoa_c + co2_c + nadh_c\n",
      "16%    0.826  MDH       mal__L_c + nad_c <=> h_c + nadh_c + oaa_c\n",
      "3%     0.149  PGCD      3pg_c + nad_c --> 3php_c + h_c + nadh_c\n",
      "None\n",
      "************************\n",
      "PRODUCING REACTIONS -- Nicotinamide adenine dinucleotide (nad_c)\n",
      "----------------------------------------------------------------\n",
      "%      FLUX  RXN ID    REACTION\n",
      "---  ------  --------  --------------------------------------------------\n",
      "59%  22.8    NADH16pp  4.0 h_c + nadh_c + q8_c --> 3.0 h_p + nad_c + q...\n",
      "38%  15      THD2pp    2.0 h_p + nadh_c + nadp_c --> 2.0 h_c + nad_c +...\n",
      "1%    0.234  HACD1     aacoa_c + h_c + nadh_c <=> 3hbcoa_c + nad_c\n",
      "1%    0.234  HACD2     3ohcoa_c + h_c + nadh_c <=> 3hhcoa_c + nad_c\n",
      "1%    0.234  HACD3     3oocoa_c + h_c + nadh_c <=> 3hocoa_c + nad_c\n",
      "1%    0.234  HACD4     3odcoa_c + h_c + nadh_c <=> 3hdcoa_c + nad_c\n",
      "1%    0.234  HACD5     3oddcoa_c + h_c + nadh_c <=> 3hddcoa_c + nad_c\n",
      "\n",
      "CONSUMING REACTIONS -- Nicotinamide adenine dinucleotide (nad_c)\n",
      "----------------------------------------------------------------\n",
      "%      FLUX  RXN ID    REACTION\n",
      "---  ------  --------  --------------------------------------------------\n",
      "47%  18.3    MDH       mal__L_c + nad_c <=> h_c + nadh_c + oaa_c\n",
      "35%  13.8    MAEA      nad_c + oaa_c --> co2_e + nadh_c + pyr_c\n",
      "13%   5.08   AKGDH     akg_c + coa_c + nad_c --> co2_c + nadh_c + succ...\n",
      "3%    1.13   PGCD      3pg_c + nad_c --> 3php_c + h_c + nadh_c\n",
      "1%    0.42   GAPD      g3p_c + nad_c + pi_c <=> 13dpg_c + h_c + nadh_c\n",
      "1%    0.296  IPMD      3c2hmp_c + nad_c --> 3c4mop_c + h_c + nadh_c\n",
      "None\n",
      "************************\n",
      "PRODUCING REACTIONS -- Nicotinamide adenine dinucleotide (nad_c)\n",
      "----------------------------------------------------------------\n",
      "%      FLUX  RXN ID    REACTION\n",
      "---  ------  --------  --------------------------------------------------\n",
      "57%  23.3    NADH16pp  4.0 h_c + nadh_c + q8_c --> 3.0 h_p + nad_c + q...\n",
      "39%  15.9    THD2pp    2.0 h_p + nadh_c + nadp_c --> 2.0 h_c + nad_c +...\n",
      "3%    1.33   GAPD      g3p_c + nad_c + pi_c <=> 13dpg_c + h_c + nadh_c\n",
      "\n",
      "CONSUMING REACTIONS -- Nicotinamide adenine dinucleotide (nad_c)\n",
      "----------------------------------------------------------------\n",
      "%      FLUX  RXN ID    REACTION\n",
      "---  ------  --------  --------------------------------------------------\n",
      "56%  22.7    MDH       mal__L_c + nad_c <=> h_c + nadh_c + oaa_c\n",
      "44%  17.7    MAEA      nad_c + oaa_c --> co2_e + nadh_c + pyr_c\n",
      "2%    0.947  PGCD      3pg_c + nad_c --> 3php_c + h_c + nadh_c\n",
      "1%    0.248  IPMD      3c2hmp_c + nad_c --> 3c4mop_c + h_c + nadh_c\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "''' look at NAD+/NADH. Would be nice to examine the predicted equilibrium ratios.'''\n",
    "print(new_cit_minus_model.metabolites.nadh_c.summary())\n",
    "print('************************')\n",
    "print(new_DM25_cit_plus_model.metabolites.nadh_c.summary())\n",
    "print('************************')\n",
    "print(new_DM0_cit_plus_model.metabolites.nadh_c.summary())\n",
    "print('************************')\n",
    "print('************************')\n",
    "print('************************')\n",
    "print(new_cit_minus_model.metabolites.nad_c.summary())\n",
    "print('************************')\n",
    "print(new_DM25_cit_plus_model.metabolites.nad_c.summary())\n",
    "print('************************')\n",
    "print(new_DM0_cit_plus_model.metabolites.nad_c.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PRODUCING REACTIONS -- Pyruvate (pyr_c)\n",
      "---------------------------------------\n",
      "%       FLUX  RXN ID      REACTION\n",
      "----  ------  ----------  --------------------------------------------------\n",
      "100%  1.58    MAEA        nad_c + oaa_c --> co2_e + nadh_c + pyr_c\n",
      "\n",
      "CONSUMING REACTIONS -- Pyruvate (pyr_c)\n",
      "---------------------------------------\n",
      "%       FLUX  RXN ID      REACTION\n",
      "----  ------  ----------  --------------------------------------------------\n",
      "83%   1.3     PDH         coa_c + nad_c + pyr_c --> accoa_c + co2_c + nadh_c\n",
      "10%   0.157   ACLS        h_c + 2.0 pyr_c --> alac__S_c + co2_c\n",
      "3%    0.0522  VPAMTr_...  3mob_c + ala__L_c <=> pyr_c + val__L_c\n",
      "2%    0.0333  DHDPS       aspsa_c + pyr_c --> 23dhdp_c + 2.0 h2o_c + h_c\n",
      "2%    0.0261  ACHBS       2obut_c + h_c + pyr_c --> 2ahbut_c + co2_c\n",
      "None\n",
      "************************\n",
      "PRODUCING REACTIONS -- Pyruvate (pyr_c)\n",
      "---------------------------------------\n",
      "%      FLUX  RXN ID      REACTION\n",
      "---  ------  ----------  --------------------------------------------------\n",
      "53%  13.8    MAEA        nad_c + oaa_c --> co2_e + nadh_c + pyr_c\n",
      "47%  12      POR5        coa_c + 2.0 flxso_c + pyr_c <=> accoa_c + co2_c...\n",
      "\n",
      "CONSUMING REACTIONS -- Pyruvate (pyr_c)\n",
      "---------------------------------------\n",
      "%      FLUX  RXN ID      REACTION\n",
      "---  ------  ----------  --------------------------------------------------\n",
      "93%  24      PFL         coa_c + pyr_c --> accoa_c + for_c\n",
      "4%    1.15   ACLS        h_c + 2.0 pyr_c --> alac__S_c + co2_c\n",
      "1%    0.382  VPAMTr_...  3mob_c + ala__L_c <=> pyr_c + val__L_c\n",
      "1%    0.243  DHDPS       aspsa_c + pyr_c --> 23dhdp_c + 2.0 h2o_c + h_c\n",
      "None\n",
      "************************\n",
      "PRODUCING REACTIONS -- Pyruvate (pyr_c)\n",
      "---------------------------------------\n",
      "%      FLUX  RXN ID      REACTION\n",
      "---  ------  ----------  --------------------------------------------------\n",
      "66%  17.7    MAEA        nad_c + oaa_c --> co2_e + nadh_c + pyr_c\n",
      "34%   9.18   POR5        coa_c + 2.0 flxso_c + pyr_c <=> accoa_c + co2_c...\n",
      "\n",
      "CONSUMING REACTIONS -- Pyruvate (pyr_c)\n",
      "---------------------------------------\n",
      "%      FLUX  RXN ID      REACTION\n",
      "---  ------  ----------  --------------------------------------------------\n",
      "93%  24.9    PFL         coa_c + pyr_c --> accoa_c + for_c\n",
      "4%    0.963  ACLS        h_c + 2.0 pyr_c --> alac__S_c + co2_c\n",
      "1%    0.34   PPS         atp_c + h2o_c + pyr_c --> amp_c + 2.0 h_c + pep...\n",
      "1%    0.32   VPAMTr_...  3mob_c + ala__L_c <=> pyr_c + val__L_c\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "''' look at pyruvate in the models'''\n",
    "print(new_cit_minus_model.metabolites.pyr_c.summary())\n",
    "print('************************')\n",
    "print(new_DM25_cit_plus_model.metabolites.pyr_c.summary())\n",
    "print('************************')\n",
    "print(new_DM0_cit_plus_model.metabolites.pyr_c.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "''' let us look for reactions which increase flux in response to higher glucose input,\n",
    "    but decrease flux in response to higher citrate input, or vice-versa.\n",
    "    These would be candidate reactions for antagonistic pleiotropy.'''\n",
    "\n",
    "## increment glucose or citrate influx from 10.0 to 11.0.\n",
    "## note that result is somewhat sensitive to +/- epsilon: MAEA is pleiotropic for -e but not +e.\n",
    "epsilon = 0.01\n",
    "\n",
    "glucose_model_dx = new_cit_minus_model.copy()\n",
    "citrate_model_dy = new_DM0_cit_plus_model.copy()\n",
    "DM25_dx = new_cit_minus_model.medium\n",
    "DM0_dy = new_DM0_cit_plus_model.medium\n",
    "DM25_dx['EX_glc__D_e'] = DM25_dx['EX_glc__D_e'] - epsilon\n",
    "DM0_dy['EX_cit_e'] = DM0_dy['EX_cit_e'] - epsilon\n",
    "glucose_model_dx.medium = DM25_dx\n",
    "citrate_model_dy.medium = DM0_dy\n",
    "glucose_dx_solution = glucose_model_dx.optimize()\n",
    "citrate_dy_solution = citrate_model_dy.optimize()\n",
    "\n",
    "citrate_solution = new_DM0_cit_plus_model.optimize()\n",
    "glucose_solution = new_cit_minus_model.optimize()\n",
    "\n",
    "dCdy = citrate_dy_solution.fluxes - citrate_solution.fluxes\n",
    "dGdx = glucose_dx_solution.fluxes - glucose_solution.fluxes\n",
    "\n",
    "flux_derivative = pd.DataFrame({'dCdy':dCdy,'dGdx':dGdx})\n",
    "flux_derivative['vec_norm'] = np.sqrt(np.square(flux_derivative).sum(axis=1))\n",
    "flux_derivative['product'] = flux_derivative['dCdy']*flux_derivative['dGdx']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "flux_derivative2 = flux_derivative[flux_derivative['vec_norm'] > epsilon]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.5686275389530429\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dCdy</th>\n",
       "      <th>dGdx</th>\n",
       "      <th>vec_norm</th>\n",
       "      <th>product</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HEX7</th>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XYLI2</th>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_glc__D_e</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLCt2pp</th>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLCtex_copy1</th>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARTRt7pp</th>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARTRtpp</th>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CITtex</th>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CITt7pp</th>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_nh4_e</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NH4tpp</th>\n",
       "      <td>-0.00</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NH4tex</th>\n",
       "      <td>-0.00</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>THD2pp</th>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPC</th>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENO</th>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGM</th>\n",
       "      <td>-0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SUCDi</th>\n",
       "      <td>-0.01</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FUM</th>\n",
       "      <td>-0.01</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACONTb</th>\n",
       "      <td>-0.01</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACONTa</th>\n",
       "      <td>-0.01</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAPD</th>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGK</th>\n",
       "      <td>-0.00</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O2tpp</th>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_o2_e</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O2tex</th>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_co2_e</th>\n",
       "      <td>-0.02</td>\n",
       "      <td>-0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PFL</th>\n",
       "      <td>-0.03</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FORtex</th>\n",
       "      <td>0.03</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_for_e</th>\n",
       "      <td>-0.03</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.03</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FORtppi</th>\n",
       "      <td>-0.03</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.03</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAEA</th>\n",
       "      <td>0.03</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>0.03</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NADH16pp</th>\n",
       "      <td>0.01</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>0.04</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CYTBO3_4pp</th>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>0.04</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CO2tpp</th>\n",
       "      <td>0.04</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CO2tex</th>\n",
       "      <td>0.04</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H2Otex</th>\n",
       "      <td>-0.02</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.06</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_h2o_e</th>\n",
       "      <td>0.02</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>0.06</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H2Otpp</th>\n",
       "      <td>-0.02</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.06</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATPS4rpp</th>\n",
       "      <td>0.02</td>\n",
       "      <td>-0.07</td>\n",
       "      <td>0.07</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FLDR2</th>\n",
       "      <td>0.28</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POR5</th>\n",
       "      <td>-0.28</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ICDHyr</th>\n",
       "      <td>0.29</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>0.29</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKGDH</th>\n",
       "      <td>0.29</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SUCOAS</th>\n",
       "      <td>-0.29</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ICL</th>\n",
       "      <td>-0.30</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MALS</th>\n",
       "      <td>-0.30</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MDH</th>\n",
       "      <td>-0.32</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPK</th>\n",
       "      <td>-0.32</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.32</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACS</th>\n",
       "      <td>0.32</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PTAr</th>\n",
       "      <td>0.32</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACKr</th>\n",
       "      <td>-0.32</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPS</th>\n",
       "      <td>0.34</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPCK</th>\n",
       "      <td>-0.34</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.34</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R15BPK</th>\n",
       "      <td>0.51</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.51</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPM</th>\n",
       "      <td>-0.51</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.51</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R1PK</th>\n",
       "      <td>0.51</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.51</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRPPS</th>\n",
       "      <td>-0.51</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>0.51</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADK1</th>\n",
       "      <td>-1.79</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>1.79</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NDPK1</th>\n",
       "      <td>1.93</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>1.93</td>\n",
       "      <td>-0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADK3</th>\n",
       "      <td>1.93</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.93</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              dCdy  dGdx  vec_norm  product\n",
       "HEX7          0.00 -0.01      0.01     -0.0\n",
       "XYLI2         0.00 -0.01      0.01     -0.0\n",
       "EX_glc__D_e   0.00  0.01      0.01      0.0\n",
       "GLCt2pp       0.00 -0.01      0.01     -0.0\n",
       "GLCtex_copy1  0.00 -0.01      0.01     -0.0\n",
       "TARTRt7pp     0.01  0.00      0.01      0.0\n",
       "TARTRtpp     -0.01  0.00      0.01     -0.0\n",
       "CITtex       -0.01  0.00      0.01     -0.0\n",
       "CITt7pp      -0.01  0.00      0.01     -0.0\n",
       "EX_nh4_e      0.00  0.01      0.01      0.0\n",
       "NH4tpp       -0.00 -0.01      0.01      0.0\n",
       "NH4tex       -0.00 -0.01      0.01      0.0\n",
       "THD2pp       -0.01  0.00      0.01     -0.0\n",
       "PPC           0.00 -0.01      0.01     -0.0\n",
       "ENO           0.00 -0.01      0.01     -0.0\n",
       "PGM          -0.00  0.01      0.01     -0.0\n",
       "SUCDi        -0.01 -0.00      0.01      0.0\n",
       "FUM          -0.01 -0.00      0.01      0.0\n",
       "ACONTb       -0.01 -0.00      0.01      0.0\n",
       "ACONTa       -0.01 -0.00      0.01      0.0\n",
       "GAPD          0.00 -0.02      0.02     -0.0\n",
       "PGK          -0.00  0.02      0.02     -0.0\n",
       "O2tpp         0.00 -0.02      0.02     -0.0\n",
       "EX_o2_e       0.00  0.02      0.02      0.0\n",
       "O2tex         0.00 -0.02      0.02     -0.0\n",
       "EX_co2_e     -0.02 -0.02      0.02      0.0\n",
       "PFL          -0.03 -0.00      0.03      0.0\n",
       "FORtex        0.03  0.00      0.03      0.0\n",
       "EX_for_e     -0.03  0.00      0.03     -0.0\n",
       "FORtppi      -0.03  0.00      0.03     -0.0\n",
       "MAEA          0.03 -0.01      0.03     -0.0\n",
       "NADH16pp      0.01 -0.04      0.04     -0.0\n",
       "CYTBO3_4pp    0.00 -0.04      0.04     -0.0\n",
       "CO2tpp        0.04  0.01      0.04      0.0\n",
       "CO2tex        0.04  0.01      0.04      0.0\n",
       "H2Otex       -0.02  0.06      0.06     -0.0\n",
       "EX_h2o_e      0.02 -0.06      0.06     -0.0\n",
       "H2Otpp       -0.02  0.06      0.06     -0.0\n",
       "ATPS4rpp      0.02 -0.07      0.07     -0.0\n",
       "FLDR2         0.28  0.00      0.28      0.0\n",
       "POR5         -0.28 -0.00      0.28      0.0\n",
       "ICDHyr        0.29 -0.00      0.29     -0.0\n",
       "AKGDH         0.29  0.00      0.29      0.0\n",
       "SUCOAS       -0.29 -0.00      0.29      0.0\n",
       "ICL          -0.30 -0.00      0.30      0.0\n",
       "MALS         -0.30 -0.00      0.30      0.0\n",
       "MDH          -0.32 -0.00      0.32      0.0\n",
       "PPK          -0.32  0.00      0.32     -0.0\n",
       "ACS           0.32  0.00      0.32      0.0\n",
       "PTAr          0.32  0.00      0.32      0.0\n",
       "ACKr         -0.32 -0.00      0.32      0.0\n",
       "PPS           0.34  0.00      0.34      0.0\n",
       "PPCK         -0.34  0.00      0.34     -0.0\n",
       "R15BPK        0.51  0.00      0.51      0.0\n",
       "PPM          -0.51  0.00      0.51     -0.0\n",
       "R1PK          0.51  0.00      0.51      0.0\n",
       "PRPPS        -0.51 -0.00      0.51      0.0\n",
       "ADK1         -1.79 -0.00      1.79      0.0\n",
       "NDPK1         1.93 -0.00      1.93     -0.0\n",
       "ADK3          1.93  0.00      1.93      0.0"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "''' this result shows no tradeoff between glucose and citrate, neglecting the costs of gene expression.\n",
    "    what if I fix oxaloacetate decarboxylation at some value? Perhaps some tradeoff. But gene regulation could\n",
    "    alleviate the tradeoff.'''\n",
    "\n",
    "''' my initial thinking/result was that maeA flux is sensitive to changes in glucose and citrate.\n",
    "    But this finding does not seem to hold up-- this idea is cool so I keep the code, but \"result\"\n",
    "    will not be presented.'''\n",
    "\n",
    "dGdx_norm = np.sqrt(np.square(flux_derivative['dGdx']).sum())\n",
    "dCdy_norm = np.sqrt(np.square(flux_derivative['dCdy']).sum())\n",
    "theta = np.arccos(flux_derivative2['dCdy'].dot(flux_derivative2['dGdx'])/(dGdx_norm*dCdy_norm))\n",
    "print(theta)\n",
    "flux_derivative2.sort_values('vec_norm').round(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
